cw/cw305/util/device.py

```diff
@@ -17,10 +44,29 @@ def __init__(self, fw_programmer, bitstream, pll_frequency, baudrate):
 
   def initialize_fpga(self, bitstream, pll_frequency):
     """Initializes FPGA bitstream and sets PLL frequency."""
-    print('Connecting and loading FPGA')
-    fpga = cw.capture.targets.CW305()
     # Do not program the FPGA if it is already programmed.
-    fpga.con(bsfile=bitstream, force=False)
+    # Note: Set this to True to force programming the FPGA when using a new
+    # bitstream.
+    # TODO: We should have this in the CLI.
+    force_programming = False
+    print('Connecting and loading FPGA... ', end = '')
+    fpga = cw.capture.targets.CW305()
+    # Runtime patch fpga.fpga.FPGAProgram to detect if it was actually called.
+    # Note: This is fragile and may break but it is easy to miss that the FPGA
```

Comment:
Maybe we can add a hard reset command to always have a deterministic way of initializing the FPGA (whenever needed).

Reply:
Yes, we should definitely expose this as a command line argument. But still, in order to print an accurate message (or to program the FPGA only if needed) we need to know the last bitstream that the FPGA was programmed with (chipwhisperer doesn't know this either AFAICT, it just checks a flag in a status packet). We can try writing the last programmed hash to a file but that would require us to introduce an on-disk cache. If we decide to do that I think this should definitely go there.

Reply:
I agree, this is an orthogonal comment. If one wants to make sure the new FPGA is loaded it can force a hard reset before running.